### ISSUE TITLE
MouseRegion no longer reports events when overflowed

### ISSUE THREAD

**Issue by luigi-rosso** [flutter/flutter#60995]
We rely on being able to place a MouseRegion anywhere in the app and detect enter/exit in that region. This helps us work around cases where the hit area needs to be different from the layout.

Is there any plan to make it toggleable to use the old hit test system or some way to permit overflow of the regions?

https://github.com/flutter/flutter/commit/6f4c4b3cf814648a644406f8d1011e7b6e0e9130#diff-429547c5504ba4c5dad4aac0d88e6c65

**Comment by goderbauer** [comment]
@luigi-rosso Do you have some short example code that shows how you were using it?

**Comment by luigi-rosso** [comment]
@goderbauer, sure! This is a trivial example that can be easily solved in other ways, but the basic concept it shows is one we rely on for more complex layout scenarios where items are tightly packed visually but still need larger hit areas. This is particularly important for desktop apps.

Example repro here: https://github.com/luigi-rosso/mouse_region_60995

Run it in **master** and notice how the MouseRegion encompassing the blue area only triggers when the darker circle has hover:
![60995_master](https://user-images.githubusercontent.com/454182/86823555-d1979180-c041-11ea-86ac-193f559a9bcf.gif)

Run it on **dev** to see how it has been working so far (since before the changes referenced above), allowing hit detection to overflow:
![60995_dev](https://user-images.githubusercontent.com/454182/86823663-fbe94f00-c041-11ea-94db-cbe5ceadd0ee.gif)

Important part that does the overflow with negative positioned items in a stack:
```dart
class _HoverCircleState extends State<HoverCircle> {
  bool _hasHover = false;
  @override
  Widget build(BuildContext context) {
    return Container(
        width: 50,
        height: 50,
        decoration: BoxDecoration(
          color: _hasHover ? Colors.red : Colors.grey,
          borderRadius: const BorderRadius.all(
            Radius.circular(25),
          ),
        ),
        child: Stack(
          overflow: Overflow.visible,
          children: [
            Positioned(
              left: -10,
              right: -10,
              top: -10,
              bottom: -10,
              child: ColoredBox(
                color: Colors.blue.withOpacity(0.25),
                child: MouseRegion(
                  opaque: true,
                  onHover: (_) {
                    setState(() {
                      _hasHover = true;
                    });
                  },
                  onExit: (_) {
                    setState(() {
                      _hasHover = false;
                    });
                  },
                ),
              ),
            ),
          ],
        ));
  }
}
```

**Comment by luigi-rosso** [comment]
We implemented a pretty kludgy workaround using Overlays. 

The workaround uses a RenderObject that pushes a layer (to help in cases where this is in a RepaintBoundary) that helps compute and cache the global bounds of the object. This gets reported up to our OverflowingMouseRegion which then places an Overlay at the global coordinates. This can now be used to include listeners/mouseregions to detect the hover/hit in this different coordinate space. It will need tweaking when content is clipped, but it unblocks us right now.

https://github.com/luigi-rosso/mouse_region_60995/compare/workaround

**Comment by dkwingsmt** [comment]
@luigi-rosso I'm afraid we're probably considering the change a fix instead of a regression, although I agree that your usage is valid and should have an easier way to achieve.

The original approach violates Flutter's assumption that a parent always encapsulate its child tree, which is used in all tree-related algorithms. This approach is very volatile with this violation and can be broken by many changes that would otherwise be safe.

Perhaps there is still a way to solve this within the framework of widgets and render objects. For example, say you have a compact row-column layout as follows:
```dart
Column(
  children: [
    EarlierRow(),
    Row(
      children: [
        EarlierCell(),
        OverflowHitTest(
          overflow: EdgeInsets.all(10),
          child: Button(),
        ),
        LaterCell(),
      ],
    ),
    LaterRow(),
  ],
)
```

We can change it to
```dart
PriorityColumn(
  children: [
    EarlierRow(),
    PriorChild(
      child: OverflowHitTest(
        overflow: EdgeInsets.fromLTRB(0, 10, 0, 10),
        child: PriorityRow(
          children: [
            EarlierCell(),
            PriorChild(
              OverflowHitTest(
                overflow: EdgeInsets.all(10)
                child: Button(),
              ),
            ),
            LaterCell(),
          ],
        ),
      ),
    )
    LaterRow(),
  ],
)
```
while introducing two mechanisms:
-  `OverflowHitTest` is a widget that only makes the hit test overflow, and does not change visual painting or constraints
-  `PriorityRow` and `PriorityColumn` are `Row` and `Column` that hit test their `PriorChild`s before others.

The biggest problem would be to implement the 2nd one. Probably make it a two-pass and do some private checks.

Of course, if the elements around the `Button` can not be hit, the 2nd will not be needed.

**Comment by luigi-rosso** [comment]
Thanks @dkwingsmt! 

This sounds great. **OverflowHitTest** sounds like exactly what we'd want and would probably work for the majority of our use cases. 

We use some tricks to do something similar to what you're doing with PriorityRow/Column. We introduced a HitDeny and HitAllow to sort of gate hit testing through the widget tree. Similar to Ignore/Absorb Pointers but they allow re-opening the hit detection set after closing it off.

**Comment by dkwingsmt** [comment]
Glad it helps! I assume we can close this issue for now? Feel free to reopen or create a new issue in case other problems occur.
